File: mobile/src/main/java/com/backdoor/moove/core/helper/SharedPrefs.java
package com.backdoor.moove.core.helper;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.Nullable;

/**
 * Helper class for working with SharedPreferences.
 */
public class SharedPrefs {

    public static final String MOOVE_PREFS = "moove_prefs";
    private static final String CHANGES_PREFS = "changes_settings";

    private SharedPreferences mPrefs;
    private SharedPreferences mChangesPrefs;
    @Nullable
    private static SharedPrefs instance;

    private SharedPrefs(Context context) {
        this.mPrefs = context.getSharedPreferences(MOOVE_PREFS, Context.MODE_PRIVATE);
        this.mChangesPrefs = context.getSharedPreferences(CHANGES_PREFS, Context.MODE_PRIVATE);
    }

    @Nullable
    public static SharedPrefs getInstance(@Nullable Context context) {
        if (instance == null && context != null) {
            synchronized (SharedPrefs.class) {
                if (instance == null) {
                    instance = new SharedPrefs(context.getApplicationContext());
                }
            }
        }
        return instance;
    }

    /**
     * Save String preference.
     *
     * @param stringToSave key.
     * @param value        value.
     */
    public void savePrefs(String stringToSave, String value) {
        mPrefs.edit().putString(stringToSave, value).apply();
    }

    /**
     * Save Integer preference.
     *
     * @param stringToSave key.
     * @param value        value.
     */
    public void saveInt(String stringToSave, int value) {
        mPrefs.edit().putInt(stringToSave, value).apply();
    }

    /**
     * Get Integer preference.
     *
     * @param stringToLoad key.
     * @return
     */
    public int loadInt(String stringToLoad) {
        try {
            return mPrefs.getInt(stringToLoad, 0);
        } catch (ClassCastException e) {
            return Integer.parseInt(mPrefs.getString(stringToLoad, "0"));
        }
    }

    /**
     * Save Long preference.
     *
     * @param stringToSave key.
     * @param value        value.
     */
    public void saveLong(String stringToSave, long value) {
        mPrefs.edit().putLong(stringToSave, value).apply();
    }

    /**
     * Get Long preference.
     *
     * @param stringToLoad key.
     * @return
     */
    public long loadLong(String stringToLoad) {
        try {
            return mPrefs.getLong(stringToLoad, 1000);
        } catch (ClassCastException e) {
            return Long.parseLong(mPrefs.getString(stringToLoad, "1000"));
        }
    }

    /**
     * Get String preference.
     *
     * @param stringToLoad key.
     * @return
     */
    public String loadPrefs(String stringToLoad) {
        try {
            return mPrefs.getString(stringToLoad, "");
        } catch (NullPointerException e) {
            return "";
        }
    }

    /**
     * Check if preference exist.
     *
     * @param checkString key.
     * @return
     */
    public boolean isString(String checkString) {
        return mPrefs.contains(checkString);
    }

    /**
     * Save Boolean preference.
     *
     * @param stringToSave key.
     * @param value        value.
     */
    public void saveBoolean(String stringToSave, boolean value) {
        mPrefs.edit().putBoolean(stringToSave, value).apply();
    }

    /**
     * Get Boolean preference.
     *
     * @param stringToLoad key.
     * @return
     */
    public boolean loadBoolean(String stringToLoad) {
        try {
            return mPrefs.getBoolean(stringToLoad, false);
        } catch (ClassCastException e) {
            return Boolean.parseBoolean(mPrefs.getString(stringToLoad, "false"));
        }
    }

    public void saveVersionBoolean(String stringToSave) {
        mChangesPrefs.edit().putBoolean(stringToSave, true).apply();
    }

    public boolean loadVersionBoolean(String stringToLoad) {
        try {
            return mChangesPrefs.getBoolean(stringToLoad, false);
        } catch (ClassCastException e) {
            return Boolean.parseBoolean(mChangesPrefs.getString(stringToLoad, "false"));
        }
    }
}

File: mobile/src/main/java/com/backdoor/moove/core/helper/SharedPrefs.kt
package com.backdoor.moove.core.helper

import android.content.Context
import android.content.SharedPreferences

/**
 * Helper class for working with SharedPreferences.
 */
class SharedPrefs private constructor(context: Context) {

    private val mPrefs: SharedPreferences
    private val mChangesPrefs: SharedPreferences

    init {
        this.mPrefs = context.getSharedPreferences(MOOVE_PREFS, Context.MODE_PRIVATE)
        this.mChangesPrefs = context.getSharedPreferences(CHANGES_PREFS, Context.MODE_PRIVATE)
    }

    /**
     * Save String preference.
     *
     * @param stringToSave key.
     * @param value        value.
     */
    fun savePrefs(stringToSave: String, value: String) {
        mPrefs.edit().putString(stringToSave, value).apply()
    }

    /**
     * Save Integer preference.
     *
     * @param stringToSave key.
     * @param value        value.
     */
    fun saveInt(stringToSave: String, value: Int) {
        mPrefs.edit().putInt(stringToSave, value).apply()
    }

    /**
     * Get Integer preference.
     *
     * @param stringToLoad key.
     * @return
     */
    fun loadInt(stringToLoad: String): Int {
        try {
            return mPrefs.getInt(stringToLoad, 0)
        } catch (e: ClassCastException) {
            return Integer.parseInt(mPrefs.getString(stringToLoad, "0")!!)
        }

    }

    /**
     * Save Long preference.
     *
     * @param stringToSave key.
     * @param value        value.
     */
    fun saveLong(stringToSave: String, value: Long) {
        mPrefs.edit().putLong(stringToSave, value).apply()
    }

    /**
     * Get Long preference.
     *
     * @param stringToLoad key.
     * @return
     */
    fun loadLong(stringToLoad: String): Long {
        try {
            return mPrefs.getLong(stringToLoad, 1000)
        } catch (e: ClassCastException) {
            return java.lang.Long.parseLong(mPrefs.getString(stringToLoad, "1000")!!)
        }

    }

    /**
     * Get String preference.
     *
     * @param stringToLoad key.
     * @return
     */
    fun loadPrefs(stringToLoad: String): String? {
        try {
            return mPrefs.getString(stringToLoad, "")
        } catch (e: NullPointerException) {
            return ""
        }

    }

    /**
     * Check if preference exist.
     *
     * @param checkString key.
     * @return
     */
    fun isString(checkString: String): Boolean {
        return mPrefs.contains(checkString)
    }

    /**
     * Save Boolean preference.
     *
     * @param stringToSave key.
     * @param value        value.
     */
    fun saveBoolean(stringToSave: String, value: Boolean) {
        mPrefs.edit().putBoolean(stringToSave, value).apply()
    }

    /**
     * Get Boolean preference.
     *
     * @param stringToLoad key.
     * @return
     */
    fun loadBoolean(stringToLoad: String): Boolean {
        try {
            return mPrefs.getBoolean(stringToLoad, false)
        } catch (e: ClassCastException) {
            return java.lang.Boolean.parseBoolean(mPrefs.getString(stringToLoad, "false"))
        }

    }

    fun saveVersionBoolean(stringToSave: String) {
        mChangesPrefs.edit().putBoolean(stringToSave, true).apply()
    }

    fun loadVersionBoolean(stringToLoad: String): Boolean {
        try {
            return mChangesPrefs.getBoolean(stringToLoad, false)
        } catch (e: ClassCastException) {
            return java.lang.Boolean.parseBoolean(mChangesPrefs.getString(stringToLoad, "false"))
        }

    }

    companion object {

        val MOOVE_PREFS = "moove_prefs"
        private val CHANGES_PREFS = "changes_settings"
        private var instance: SharedPrefs? = null

        fun getInstance(context: Context?): SharedPrefs? {
            if (instance == null && context != null) {
                synchronized(SharedPrefs::class.java) {
                    if (instance == null) {
                        instance = SharedPrefs(context.applicationContext)
                    }
                }
            }
            return instance
        }
    }
}